File: frontend/gameapp-frontend/src/domain/value-objects/scenes/enter-dungeon-scene.js
import { Scene } from "../../entities/scene";
import { ChangeBiomeScene } from "./change-biome-scene";

// EnterDungeonScene
export class EnterDungeonScene extends Scene {
    constructor(name, description, biome, possibleScenes = [], id = null) {
        super(name, description, biome, id);
        this._possibleScenes = possibleScenes;
    }

    // getter
    get possibleScenes() { return [...this._possibleScenes]; }

    // setter
    setSceneName(newName) { return new ChangeBiomeScene(newName, this._description, this._biome, this._possibleScenes); }
    setSceneDescription(newDescription) { return new ChangeBiomeScene(this._name, newDescription, this._biome, this._possibleScenes); }
    setBiome(newBiome) { return new ChangeBiomeScene(this._name, this._description, newBiome, this._possibleScenes); }


    setPossibleScenes(newScenes) {
        return new EnterDungeonScene(this._name, this._description, this._biome, newScenes);
    }

    generateRandomScene() {
        const randomIndex = Math.floor(Math.random() * this._possibleScenes.length);
        return this._possibleScenes[randomIndex];
    }

    toString() {
        const scenesStr = this._possibleScenes.map(s => s.name.toString()).join(", ");
        return `EnterDungeonScene: ${this._name} - ${this._biome}\nDescription: ${this._description}\nPossible Scenes: [${scenesStr}]`;
    }
}

File: backend/GameApp.Domain/Entities/Scenes/EnterDungeonScene.cs
using GameApp.Domain.Enumerates;
using GameApp.Domain.ValueObjects.Scenes;

namespace GameApp.Domain.Entities.Scenes;

// class for a scene where you enter a dungeon that can randomly generate a scene
public class EnterDungeonScene : Scene
{

    private static readonly Random _random = new Random();
    private readonly List<Scene> PossibleScenes;

    // Constructor
    public EnterDungeonScene(SceneName name, SceneDescription description, Biome biome, List<Scene> possibleScenes)
    : base(name, description, biome)
    {
        PossibleScenes = possibleScenes;
    }

    // Restore Constructor
    public EnterDungeonScene(Guid id, SceneName name, SceneDescription description, Biome biome, List<Scene> possibleScenes)
    : base(id, name, description, biome)
    {
        PossibleScenes = possibleScenes;
    }

    // Getter
    public List<Scene> GetPossibleScenes() => new List<Scene>(PossibleScenes);

    // Setter
    public EnterDungeonScene SetSceneName(SceneName newName) =>
    new EnterDungeonScene(newName, GetDescription(), GetBiome(), PossibleScenes);

    public EnterDungeonScene SetSceneDescription(SceneDescription newDescription) =>
    new EnterDungeonScene(GetName(), newDescription, GetBiome(), PossibleScenes);

    public EnterDungeonScene SetBiome(Biome newBiome) =>
    new EnterDungeonScene(GetName(), GetDescription(), newBiome, PossibleScenes);

    public EnterDungeonScene SetPossibleScenes(List<Scene> newPossibleScenes) => new EnterDungeonScene(GetName(), GetDescription(), GetBiome(), newPossibleScenes);

    // Generate Random Scene
    public Scene GenerateRandomScene()
    {
        int randomPosScene = _random.Next(PossibleScenes.Count);

        Scene scene = PossibleScenes[randomPosScene];

        return scene;
    }

    // To string
    public override string ToString()
    {
        string scenes = string.Join(", ", PossibleScenes.Select(s => s.GetName().ToString()));
        return $"EnterDungeonScene({GetGuid()}): {GetName()} - {GetBiome()}\nDescription: {GetDescription()}\nPossible Scenes: [{scenes}]";
    }

}